File: 0045-jump-game-ii/0045-jump-game-ii.java
class Solution {
    public int jump(int[] nums) {
        int MAX_VAL = (int) (1e4 + 1);
        int[] dp = new int[nums.length];
        Arrays.fill(dp, MAX_VAL);
        dp[nums.length - 1] = 0;
        for (int i = nums.length - 2; i >= 0; i--) {
            int jumps = nums[i];
            for (int j = 1; j <= jumps && i + j < nums.length; j++) {
                dp[i] = Math.min(dp[i], 1 + dp[i + j]);
            }
        }
        return dp[0];
    }
}

File: 0045-jump-game-ii/0045-jump-game-ii.swift
class Solution {

func jump(_ nums: [Int]) -> Int {
    var memo: [Int] = []
    for i in 0...nums.count - 1 {
        memo.append(Int.max)
    }
    return helper(nums, at: 0, jumps: 0, memo: &memo)
}

func helper(_ nums: [Int], at index: Int, jumps: Int, memo: inout [Int]) -> Int {
    if index == nums.count - 1 {
        return jumps
    }
    var ans = Int.max
    if (nums[index] == 0) {
        return ans
    }
    if memo[index] != Int.max {
        return memo[index] + jumps
    }
    for i in 1...nums[index] {
        if index + i < nums.count {
            ans = min(ans, helper(nums, at: index + i, jumps: jumps + 1, memo: &memo))
        } else {
            break
        }
    }
    memo[index] = ans - jumps
    return ans
}
}